prog.3
.\" Copyright 1999 raf (raf2@zip.com.au)
.TH prog 3  "25 June 1999" "raf" "Twisted Systems Freeware"
.SH NAME
prog_set_name,
prog_set_usage,
prog_set_desc,
prog_set_version,
prog_set_date,
prog_set_author,
prog_set_contact,
prog_set_vendor,
prog_set_url,
prog_set_legal,
prog_set_msg,
prog_set_dbg,
prog_set_debug_level,
prog_name,
prog_usage,
prog_desc,
prog_version,
prog_date,
prog_author,
prog_contact,
prog_vendor,
prog_url,
prog_legal,
prog_msg,
prog_dbg,
prog_debug_level,
prog_usage_msg,
prog_help_msg,
prog_help_check,
prog_debug_check,
prog_remove_arg,
prog_basename,
prog_error_stderr,
prog_error_file,
prog_error_syslog,
prog_error_none,
prog_debug_stderr,
prog_debug_file,
prog_debug_syslog,
prog_debug_none
\-
general program utilities: identification, debug/error/log messages.
.SH SYNOPSIS
.nf
.B #include <daemon/prog.h>
.sp
.BI "const char *prog_set_name(const char *" name )
.sp
.BI "const char *prog_set_usage(const char *" usage )
.sp
.BI "const char *prog_set_desc(const char *" desc )
.sp
.BI "const char *prog_set_version(const char *" version )
.sp
.BI "const char *prog_set_date(const char *" date )
.sp
.BI "const char *prog_set_author(const char *" author )
.sp
.BI "const char *prog_set_vendor(const char *" vendor )
.sp
.BI "const char *prog_set_contact(const char *" contact )
.sp
.BI "const char *prog_set_url(const char *" url )
.sp
.BI "const char *prog_set_legal(const char *" legal )
.sp
.BI "Msg *prog_set_msg(Msg *" msg )
.sp
.BI "Msg *prog_set_dbg(Msg *" dbg )
.sp
.BI "size_t prog_set_debug_level(size_t " level )
.sp
.BI "const char *prog_name(void)"
.sp
.BI "const char *prog_usage(void)"
.sp
.BI "const char *prog_desc(void)"
.sp
.BI "const char *prog_version(void)"
.sp
.BI "const char *prog_date(void)"
.sp
.BI "const char *prog_author(void)"
.sp
.BI "const char *prog_vendor(void)"
.sp
.BI "const char *prog_contact(void)"
.sp
.BI "const char *prog_url(void)"
.sp
.BI "const char *prog_legal(void)"
.sp
.BI "Msg *prog_msg(void)"
.sp
.BI "Msg *prog_dbg(void)"
.sp
.BI "size_t prog_debug_level(void)"
.sp
.BI "void prog_usage_msg(const char *" fmt ", ...)"
.sp
.BI "void prog_help_msg(void)"
.sp
.BI "void prog_help_check(int " ac ", char **" av )
.sp
.BI "void prog_debug_check(int *" ac ", char ***" av )
.sp
.BI "void prog_remove_arg(int *" ac ", char ***" av ", int " a )
.sp
.BI "char *prog_basename(const char *" path )
.sp
.BI "int prog_error_stderr(void)"
.sp
.BI "int prog_error_logfile(const char *" path )
.sp
.BI "int prog_error_syslog(const char *" ident ", int " option ", int " facility )
.sp
.BI "int prog_error_none(void)"
.sp
.BI "int prog_debug_stderr(void)"
.sp
.BI "int prog_debug_logfile(const char *" path )
.sp
.BI "int prog_debug_syslog(const char *" id ", int " option ", int " facility )
.sp
.BI "int prog_debug_none(void)"
.fi
.SH DESCRIPTION
.I debug()
outputs a debug message if
.I lvl
is less than or equal to the program's current debug level.
.I fmt
is a printf-like format string and processes any remaining arguments
in the same way as printf.
.PP
.I error()
outputs an error message. If the program's name has been supplied using
prog_set_name(), the error message will be preceeded by the name, a colon
and a space.
.I fmt
is a printf-like format string and processes any remaining arguments in the
same way as printf. Returns -1.
.PP
.I fatal()
outputs an error message and exits with a return code of 1. The error
message, will contain the string "fatal: ".
.I fmt
is a printf-like format string and processes any remaining arguments
in the same way as printf.
.PP
.I dump()
outputs an error message and then abort()s the program. The error message,
will contain the string "dump: ".
.I fmt
is a printf-like format string and processes any remaining arguments in the
same way as printf.
.PP
.I debugsys()
outputs a debug message if
.I lvl
is less than or equal to the program's current debug level.
.I fmt
is a printf-like format string and processes any remaining arguments in the
same way as printf. After the message is a colon, a space and the string
version of the current value of errno.
.PP
.I errorsys()
outputs an error message. If the program's name has been supplied using
prog_set_name(), the error message will be preceeded by the name, a colon
and a space.
.I fmt
is a printf-like format string and processes any remaining arguments in the
same way as printf. After the message is a colon, a space and the string
version of the current value of errno. Returns -1.
.PP
.I fatalsys()
outputs an error message and exits with a return code of 1. The error
message, will contain the string "fatal: ".
.I fmt
is a printf-like format string and processes any remaining arguments in the
same way as printf. After the message is a colon, a space and the string
version of the current value of errno.
.PP
.I dumpsys()
outputs an error message and then abort()s the program. The error message,
will contain the string "dump: ".
.I fmt
is a printf-like format string and processes any remaining arguments in the
same way as printf. After the message is a colon, a space and the string
version of the current value of errno.
.PP
.I errno_set()
sets errno to errnum and returns -1.
.SH "SEE ALSO"
.BR libdaemon(3)


CHANGELOG
CHANGELOG
=========

0.8.1 (20230313)

    - Fixed compiler error when NDEBUG is defined (Baruch Siach)
    - When NDEBUG is defined exclude more unused code (fix warning, smaller binary)
    - Fixed buglet: Only delete the pidfile/clientpidfile we created (James Abbatiello abbeyj at gmail.com)
    - Added CODE_OF_CONDUCT.md
    - Add LICENSES/* and .reuse/dep5 (REUSE 3.0 compliant)
    - Add references to the codeberg repository

0.8 (20210304)

    - Fixed formatting typo in README.md
    - Added support for (Linux only) musl libc by including <sys/ttydefaults.h> for CEOF (by Baruch Siach)
    - Added --bind option to terminate when the user logs out (iff systemd/elogind) (with ZhaoLin1457 et al.)
    - Added Slackware-specific configuration
    - Added expansion of $VAR and ~user notation in command line option arguments and configuration files
    - Added environment variable definitions in configuration files
    - Automatically create --pidfiles or --pidfile directory if it's under the user's home directory
    - Updated README.md (improved notes on --stop and --restart, and added --signal --running --list --bind)
    - Added configure --help, --destdir, --enable-logind and --disable-logind options
    - Fixed $(DESTDIR) use in makefiles to interoperate with configure --prefix

0.7.1 (20210220)

    - Fixed make distclean/clobber on Debian (use debian-clobber instead of clobber) (spotted by Marc Haber)
    - Fixed parallel make test (race condition creating test directory) (spotted by Marc Haber)
    - Fixed manpage header and footer and various other minor fixes and improvements
    - Minor improvements in a few module tests
    - Added configure --prefix=/opt/local for macports (spotted by Kurt Hindenburg)
    - Added mention of file locations on macOS/macports to the daemon(1) manpage
    - Added checks for gcc -W options that fail on old macosx (e.g. 10.6.8) without macports
    - Minor changes to usage error messages for consistency
    - Improved config file parsing (skip spaces around commas and added more error messages)
    - Added support for using $CC in configure (spotted by Ryan Schmidt)
    - Fixed Makefile (wasn't inheriting CPPFLAGS, CFLAGS or LDFLAGS from caller) (spotted by Ryan Schmidt)

0.7 (20201111)

    - Fixed bug: race condition causing failure with "exec daemon ..." (with Askar Safin safinaskar at mail.ru)
    - Added $DAEMON_INIT_EXIT_DELAY_MSEC workaround for obscure KDE/konsole bug (with Askar Safin safinaskar at mail.ru)
    - Removed O_NOCTTY-related manpage "bug" (now historical) (spotted by Askar Safin safinaskar at mail.ru)
    - Always double fork for safety on systems of unknown flavour (suggested by Askar Safin safinaskar at mail.ru)
    - Added support for GNU/Hurd (by Svante Signell svante.signell at telia.com)
    - Added --ignore-eof and --read-eof options to alter behaviour when client dies without closing stdio
      (for Terry Greeniaus greeniaus_terry at emc.com and Tilman Baumann at tilman.baumann at grandeye.com)
    - Fixed bug: daemon_lock_pidfile() leaked fd across exec (spotted by Tilman Baumann tilman.baumann at grandeye.com)
    - Added --signal option (suggested by Marc Haber mh+daemon-dev at zugschlus.de)
    - Added --nocore option and removed silly manpage "bug" about its absence
    - Added fcntl_set_fdflag() and fcntl_clear_fdflag() to the fio module
    - Added signal_set_siginfo_handler() to the sig module
    - Added msg_filter_t, msg_create_filter(), msg_create_filter_with_locker() to the msg module
    - Test secure memory in the mem module for non-root users as well (now that they can do it)
    - Added prog_{out,err,dbg,alert}_push_filter() to the prog module
    - Added recvcred(), recvfromcred() to the net module
    - Added a few more debug messages that were missing
    - Insert daemon name before client's argv[0] so it appears in ps output (for marko.asplund at gmail.com)
    - Allow --pidfiles to be present without --name also being present
    - Added --list option to print a list of named daemons (more detail with --verbose)
    - Improved error message when client exit()s after --stop (for jakubholy.net at gmail.com)
    - Added syslog facility and priority lists to daemon(1) manpage (for jakubholy.net at gmail.com)
    - Various fixes and improvements to manpage
    - Improved --verbose --running message to state if client is running as well or not (added client pid file)
    - Fixed buglet: daemon no longer exits with EXIT_FAILURE when client is killed by an unexpected signal
    - Fixed harmless format string bug in prog_opt_process() (spotted by Jong-Gwon Kim and Woosuk Lee {jgkim,wslee} at ropas.snu.ac.kr)
    - Fixed bug: leaked fds for --errlog, --dbglog, --stdout and --stderr across exec
    - Fixed bug in parallelized make: "make ready" first so do it in configure (spotted by ilovezfs at icloud.com)
    - Fixed copyright messages (list actual years of publication, not just a range of years)
    - Fixed all sscanf "%s" to limit output (e.g. "%63s") to prevent buffer overruns (was only in test and example code)
    - Added more tests in the prop module
    - Fixed buglet: Removed vhangup() in pty_make_controlling_tty() as it broke some coproc module tests if run as root
    - Fixed str.h to not clash with new strlcpy/strlcat macros on OSX/macOS
    - Updated to avoid new warnings in recent versions of gcc
    - Split README into multiple files (README.md INSTALL COPYING REFERENCES CHANGELOG)
    - Added /etc/daemon.conf.d/ and ~/.daemonrc.d/ configuration directories
    - Changed "config" script name to "configure" (not an alias anymore) (but it's still not GNU configure)
    - Changed /etc/daemon.conf file location to /usr/local/etc/daemon.conf on FreeBSD
    - Changed /etc/daemon.conf file location to /usr/pkg/etc/daemon.conf on NetBSD
    - Changed installation location prefix to /usr/pkg on NetBSD
    - Added concrete examples to daemon.conf

prog.c
/*
* daemon: http://www.zip.com.au/~raf2/lib/software/daemon
*
* Copyright (c) 1999 raf
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#include <unistd.h>

#include "msg.h"
#include "err.h"
#include "prog.h"
#include "opt.h"

#ifdef NEEDS_SNPRINTF
#include "snprintf.h"
#endif

typedef struct Prog Prog;

struct Prog
{
	const char *name;
	Options *options;
	const char *syntax;
	const char *desc;
	const char *version;
	const char *date;
	const char *author;
	const char *contact;
	const char *vendor;
	const char *url;
	const char *legal;
	Msg *out;
	Msg *err;
	Msg *dbg;
	size_t level;
};

static Prog g =
{
	NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
	NULL, NULL, NULL, 0
};

/*
** void prog_init(void)
**
** Initialises the message, error and debug destinations.
*/

void prog_init(void)
{
	prog_out_stdout();
	prog_err_stderr();
	prog_dbg_stderr();
}

/*
** const char *prog_set_name(const char *name)
**
** Sets the program's name to the string name.
*/

const char *prog_set_name(const char *name)
{
	return g.name = name;
}

/*
** Options *prog_set_options(Options *options)
**
** Sets the program's options to usage.
*/

Options *prog_set_options(Options *options)
{
	return g.options = options;
}

/*
** const char *prog_set_syntax(const char *syntax)
**
** Sets the program's syntax description.
*/

const char *prog_set_syntax(const char *syntax)
{
	return g.syntax = syntax;
}

/*
** const char *prog_set_desc(const char *desc)
**
** Sets the program's description to the string desc.
*/

const char *prog_set_desc(const char *desc)
{
	return g.desc = desc;
}

/*
** const char *prog_set_version(const char *version)
**
** Sets the program's version to the string version.
*/

const char *prog_set_version(const char *version)
{
	return g.version = version;
}

/*
** const char *prog_set_date(const char *date)
**
** Sets the program's release date to the string date.
*/

const char *prog_set_date(const char *date)
{
	return g.date = date;
}

/*
** const char *prog_set_author(const char *author)
**
** Sets the program's author to the string author.
*/

const char *prog_set_author(const char *author)
{
	return g.author = author;
}

/*
** const char *prog_set_contact(const char *contact)
**
** Sets the program's contact address to the string contact.
*/

const char *prog_set_contact(const char *contact)
{
	return g.contact = contact;
}

/*
** const char *prog_set_vendor(const char *vendor)
**
** Sets the program's vendor to the string vendor.
*/

const char *prog_set_vendor(const char *vendor)
{
	return g.vendor = vendor;
}

/*
** const char *prog_set_url(const char *url)
**
** Sets the program's URL to the string url.
*/

const char *prog_set_url(const char *url)
{
	return g.url = url;
}

/*
** const char *prog_set_legal(const char *legal)
**
** Sets the program's legal notice to the string legal.
*/

const char *prog_set_legal(const char *legal)
{
	return g.legal = legal;
}

/*
** Msg *prog_set_out(Msg *output)
**
** Sets the program's message destination to out.
*/

Msg *prog_set_out(Msg *out)
{
	if (g.out && g.out != out)
		msg_release(g.out);

	return g.out = out;
}

/*
** Msg *prog_set_err(Msg *err)
**
** Sets the program's error message destination to msg.
*/

Msg *prog_set_err(Msg *err)
{
	if (g.err && g.err != err)
		msg_release(g.err);

	return g.err = err;
}

/*
** Msg *prog_set_dbg(Msg *dbg)
**
** Sets the program's debug message destination to dbg.
*/

Msg *prog_set_dbg(Msg *dbg)
{
	if (g.dbg && g.dbg != dbg)
		msg_release(g.dbg);

	return g.dbg = dbg;
}

/*
** size_t prog_set_debug_level(size_t level)
**
** Sets the program's debug level to level. Returns the previous debug level.
*/

size_t prog_set_debug_level(size_t level)
{
	size_t prev = g.level;
	g.level = level;
	return prev;
}

/*
** const char *prog_name(void)
**
** Returns the program's name.
*/

const char *prog_name(void)
{
	return g.name;
}

/*
** const Options *prog_options(void)
**
** Returns the program's options.
*/

const Options *prog_options(void)
{
	return g.options;
}

/*
** const char *prog_syntax(void)
**
** Returns the program's syntax description.
*/

const char *prog_syntax(void)
{
	return g.syntax;
}

/*
** const char *prog_desc(void)
**
** Returns the program's description.
*/

const char *prog_desc(void)
{
	return g.desc;
}

/*
** const char *prog_version(void)
**
** Returns the program's version string.
*/

const char *prog_version(void)
{
	return g.version;
}

/*
** const char *prog_date(void)
**
** Returns the program's release date.
*/

const char *prog_date(void)
{
	return g.date;
}

/*
** const char *prog_author(void)
**
** Returns the program's author.
*/

const char *prog_author(void)
{
	return g.author;
}

/*
** const char *prog_contact(void)
**
** Returns the program's contact address.
*/

const char *prog_contact(void)
{
	return g.contact;
}

/*
** const char *prog_vendor(void)
**
** Returns the program's vendor.
*/

const char *prog_vendor(void)
{
	return g.vendor;
}

/*
** const char *prog_url(void)
**
** Returns the program's URl.
*/

const char *prog_url(void)
{
	return g.url;
}

/*
** const char *prog_legal(void)
**
** Returns the program's legal notice.
*/

const char *prog_legal(void)
{
	return g.legal;
}

/*
** Msg *prog_out(void)
**
** Returns the program's message destination.
*/

Msg *prog_out(void)
{
	return g.out;
}

/*
** Msg *prog_err(void)
**
** Returns the program's error message destination.
*/

Msg *prog_err(void)
{
	return g.err;
}

/*
** Msg *prog_dbg(void)
**
** Returns the program's debug message destination.
*/

Msg *prog_dbg(void)
{
	return g.dbg;
}

/*
** size_t prog_debug_level(void)
**
** Returns the program's debug level.
*/

size_t prog_debug_level(void)
{
	return g.level;
}

/*
** int prog_out_fd(int fd)
**
** Sets the program's message destination to be file descriptor
** specified by fd. Returns 0 on success, -1 on error.
*/

int prog_out_fd(int fd)
{
	Msg *msg;

	if (!(msg = msg_create_fd(fd)))
		return -1;

	prog_set_out(msg);

	return 0;
}

/*
** int prog_out_stdout(void)
**
** Sets the program's message destination to be standard output.
** Returns 0 on success, -1 on error.
*/

int prog_out_stdout(void)
{
	return prog_out_fd(STDOUT_FILENO);
}

/*
** int prog_out_file(const char *path)
**
** Sets the program's message destination to be the file
** specified by path. Returns 0 on success, -1 on error.
*/

int prog_out_file(const char *path)
{
	Msg *msg;

	if (!(msg = msg_create_file(path)))
		return -1;

	prog_set_out(msg);

	return 0;
}

/*
** int prog_out_syslog(const char *ident, int option, int facility)
**
** Sets the program's message destination to be the system logger
** initialised with ident, option and facility.
** Returns 0 on success, -1 on error.
*/

int prog_out_syslog(const char *ident, int option, int facility)
{
	Msg *msg;

	if (!(msg = msg_create_syslog(ident, option, facility)))
		return -1;

	prog_set_out(msg);

	return 0;
}

/*
** int prog_out_none(void)
**
** Sets the program's message destination to be NULL. This disables
** all error messages. Returns 0.
*/

int prog_out_none(void)
{
	prog_set_out(NULL);

	return 0;
}

/*
** int prog_err_fd(int fd)
**
** Sets the program's error message destination to be file descriptor
** specified by fd. Returns 0 on success, -1 on error.
*/

int prog_err_fd(int fd)
{
	Msg *msg;

	if (!(msg = msg_create_fd(fd)))
		return -1;

	prog_set_err(msg);

	return 0;
}

/*
** int prog_err_stderr(void)
**
** Sets the program's error message destination to be standard error.
** Returns 0 on success, -1 on error.
*/

int prog_err_stderr(void)
{
	return prog_err_fd(STDERR_FILENO);
}

/*
** int prog_err_file(const char *path)
**
** Sets the program's error message destination to be the file
** specified by path. Returns 0 on success, -1 on error.
*/

int prog_err_file(const char *path)
{
	Msg *msg;

	if (!(msg = msg_create_file(path)))
		return -1;

	prog_set_err(msg);

	return 0;
}

/*
** int prog_err_syslog(const char *ident, int option, int facility)
**
** Sets the program's error message destination to be the system logger
** initialised with ident, option and facility.
** Returns 0 on success, -1 on error.
*/

int prog_err_syslog(const char *ident, int option, int facility)
{
	Msg *msg;

	if (!(msg = msg_create_syslog(ident, option, facility)))
		return -1;

	prog_set_err(msg);

	return 0;
}

/*
** int prog_err_none(void)
**
** Sets the program's error message destination to be NULL. This disables
** all error messages. Returns 0.
*/

int prog_err_none(void)
{
	prog_set_err(NULL);

	return 0;
}

/*
** int prog_dbg_fd(int fd)
**
** Sets the program's debug message destination to be file descriptor
** specified by fd. Returns 0 on success, -1 on error.
*/

int prog_dbg_fd(int fd)
{
	Msg *msg;

	if (!(msg = msg_create_fd(fd)))
		return -1;

	prog_set_dbg(msg);

	return 0;
}

/*
** int prog_dbg_stdout(void)
**
** Sets the program's debug message destination to be standard output.
** Returns 0 on success, -1 on error.
*/

int prog_dbg_stdout(void)
{
	return prog_dbg_fd(STDOUT_FILENO);
}

/*
** int prog_dbg_stderr(void)
**
** Sets the program's debug message destination to be standard error.
** Returns 0 on success, -1 on error.
*/

int prog_dbg_stderr(void)
{
	return prog_dbg_fd(STDERR_FILENO);
}

/*
** int prog_dbg_file(const char *path)
**
** Sets the program's debug message destination to be the file
** specified by path.
*/

int prog_dbg_file(const char *path)
{
	Msg *dbg;

	if (!(dbg = msg_create_file(path)))
		return -1;

	prog_set_dbg(dbg);

	return 0;
}

/*
** int prog_dbg_syslog(const char *id, int option, int facility)
**
** Sets the program's debug message destination to be the system logger
** initialised with ident, option and facility.
*/

int prog_dbg_syslog(const char *id, int option, int facility)
{
	Msg *dbg;

	if (!(dbg = msg_create_syslog(id, option, facility)))
		return -1;

	prog_set_dbg(dbg);

	return 0;
}

/*
** int prog_dbg_none(void)
**
** Sets the program's debug message destination to be NULL. This disables
** all debug messages. Returns 0.
*/

int prog_dbg_none(void)
{
	prog_set_dbg(NULL);

	return 0;
}

/*
** int prog_opt_process(int ac, const char **av)
**
** Parses the command line arguments in argv to process options.
*/

int prog_opt_process(int ac, char **av)
{
	int rc = opt_process(ac, av, g.options);
	if (rc == -1)
		prog_usage_msg("");

	return rc;
}

/*
** void prog_usage_msg(const char *fmt, ...)
**
** Emits a program usage error message. fmt is a printf-like format string.
** Any remaining arguments are processed as in printf(2).
*/

void prog_usage_msg(const char *fmt, ...)
{
	char msg_buf[BUFSIZ];
	char opt_buf[BUFSIZ];
	va_list args;
	va_start(args, fmt);
	vsnprintf(msg_buf, BUFSIZ, fmt, args);
	va_end(args);
	opt_usage(opt_buf, BUFSIZ, g.options);

	fflush(stderr); /* I thought stderr wasn't supposed to be buffered? */

	msg_out(g.err, "%s%susage: %s %s\n%s",
		msg_buf,
		strlen(msg_buf) ? "\n" : "",
		(g.name) ? g.name : "",
		(g.syntax) ? g.syntax : "",
		opt_buf
	);

	exit(1);
}

/*
** void prog_help_msg(void)
**
** Emits a program help message. This message consists of the program's usage
** message, description, version, release date, author, URL and legal notice.
*/

static void prog_help_msg(void)
{
#define safe(str) ((str) ? (str) : "")
	char buf[BUFSIZ];
	size_t length = 0;

	snprintf(buf, BUFSIZ, "usage: %s %s\n", safe(g.name), safe(g.syntax));

	if (g.options)
	{
		length = strlen(buf);
		opt_usage(buf + length, BUFSIZ - length, g.options);
	}

	if (g.desc)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "%s\n", g.desc);
	 }

	if (g.version)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "Version: %s\n", g.version);
	}

	if (g.date)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "Date: %s\n", g.date);
	}

	if (g.author)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "Author: %s\n", g.author);
	}

	if (g.contact)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "Contact: %s\n", g.contact);
	}

	if (g.vendor)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "Vendor: %s\n", g.vendor);
	}

	if (g.url)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "URL: %s\n", g.url);
	}

	if (g.legal)
	{
		length = strlen(buf);
		snprintf(buf + length, BUFSIZ - length, "\n%s\n", g.legal);
	}

	msg("%s", buf);
	exit(0);
}

/*
** void prog_version_msg(void)
**
** Emits a program version message. This message consists of the program's
** name and version.
*/

static void prog_version_msg(void)
{
	char buf[BUFSIZ];

	if (g.name && g.version)
		snprintf(buf, BUFSIZ, "%s-%s\n", g.name, g.version);
	else if (g.name)
		snprintf(buf, BUFSIZ, "%s\n", g.name);
	else if (g.version)
		snprintf(buf, BUFSIZ, "%s\n", g.version);
	else
		*buf = '\0';

	msg("%s", buf);
	exit(0);
}

/*
** const char *prog_basename(const char *path)
**
** Returns the filename part of path.
*/

const char *prog_basename(const char *path)
{
	const char *name = strrchr(path, PATH_SEP);

	return (name) ? (name + 1) : (char*)path;
}

static Option prog_optab[] =
{
	{ "help", NULL, "Prints a help message then exits", no_argument, OPT_NONE, OPT_FUNCTION, (void*)prog_help_msg },
	{ "version", NULL, "Prints a version message then exits", no_argument, OPT_NONE, OPT_FUNCTION, (void*)prog_version_msg },
	{ "debug", "level", "Sets the debug level", optional_argument, OPT_NONE, OPT_FUNCTION, (void*)prog_set_debug_level },
	{ NULL, NULL, NULL, 0, 0, 0, NULL }
};

Options prog_options_table[1] = {{ NULL, 0, prog_optab }};

#ifdef TEST

#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/wait.h>

int verify(int i, const char *name, const char *result, const char *prog_name, const char *type)
{
	char buf[BUFSIZ];
	char result_buf[BUFSIZ];
	int fd;
	ssize_t bytes;

	if ((fd = open(name, O_RDONLY)) == -1)
	{
		printf("Test%d: failed to output message\n", i);
		return 1;
	}

	memset(buf, '\0', BUFSIZ);
	bytes = read(fd, buf, BUFSIZ);
	close(fd);
	unlink(name);

	if (bytes == -1)
	{
		printf("Test%d: failed to read output (%s)\n", i, strerror(errno));
		return 1;
	}

	snprintf(result_buf, BUFSIZ, result, prog_name, prog_name);

	if (strcmp(buf, result_buf))
	{
		printf("Test%d: incorrect output:\nshould be:\n%s\nwas:\n%s\n", i, result_buf, buf);
		return 1;
	}

	return 0;
}

int main(int ac, char **av)
{
	const char *prog_name;
	char out_name[16];
	char err_name[16];
	int out, err;
	int tests = 3;
	int argc = 3;
	char *argv[3][4] =
	{
		{ NULL, "-debug=4", "-help", NULL },
		{ NULL, "-debug", "-version", NULL },
		{ NULL, "-debug", "-invalid", NULL }
	};
	char *results[3][2] =
	{
		/* -help output */
		{
			/* stdout */
			"usage: %s [options]\n"
			"        -help          -- Prints a help message then exits\n"
			"        -version       -- Prints a version message then exits\n"
			"        -debug [level] -- Sets the debug level\n"
			"This program tests the prog module.\n"
			"Version: 0.1\n"
			"Date: 13/10/1999\n"
			"Author: raf <raf2@zip.com.au>\n"
			"Contact: raf <raf2@zip.com.au>\n"
			"Vendor: Twisted Systems\n"
			"URL: http://www.zip.com.au/~raf2/lib/software/daemon\n"
			"\n"
			"This software is free. Use it at your own risk.\n",

			/* stderr */
			""
		},

		/* -version output */

		{
			/* stdout */
			"%s-0.1\n",

			/* stderr */
			""
		},

		/* -invalid output */
		{
			/* stdout */
			"",

			/* stderr */
			"%s: unrecognized option `-invalid'\n"
			"usage: %s [options]\n"
			"        -help          -- Prints a help message then exits\n"
			"        -version       -- Prints a version message then exits\n"
			"        -debug [level] -- Sets the debug level\n"
		}
	};

	int errors = 0;
	int i;

	printf("Testing: %s\n", *av);

	prog_name = prog_basename(*av);
	argv[0][0] = argv[1][0] = argv[2][0] = (char*)prog_name;

	prog_init();
	prog_set_name(prog_name);
	prog_set_syntax("[options]");
	prog_set_options(prog_options_table);
	prog_set_version("0.1");
	prog_set_date("13/10/1999");
	prog_set_author("raf <raf2@zip.com.au>");
	prog_set_contact("raf <raf2@zip.com.au>");
	prog_set_vendor("Twisted Systems");
	prog_set_url("http://www.zip.com.au/~raf2/lib/software/daemon");
	prog_set_legal("This software is free. Use it at your own risk.");
	prog_set_desc("This program tests the prog module.");

	out = dup(STDOUT_FILENO);
	err = dup(STDERR_FILENO);

	for (i = 0; i < tests; ++i)
	{
		pid_t pid = fork();

		snprintf(out_name, 16, "./prog.out.%d", i);
		snprintf(err_name, 16, "./prog.err.%d", i);
		freopen(out_name, "a", stdout);
		freopen(err_name, "a", stderr);

		switch (pid)
		{
			case 0:
			{
				return (prog_opt_process(argc, argv[i]) != argc);
			}

			case -1:
			{
				dup2(out, STDOUT_FILENO);
				dup2(err, STDERR_FILENO);
				++errors, printf("Test%d: failed to perform test - fork() failed (%s)\n", i + 1, strerror(errno));
				continue;
			}

			default:
			{
				int status[1];

				dup2(out, STDOUT_FILENO);
				dup2(err, STDERR_FILENO);

				if (waitpid(pid, status, 0) == -1)
				{
					++errors, printf("Test%d: failed to wait for test - waitpid(%d) failed (%s)\n", i + 1, pid, strerror(errno));
					continue;
				}

				if (WIFSIGNALED(*status))
					++errors, printf("Test%d: failed: received signal %d\n", i + 1, WTERMSIG(*status));

				if (i != 2 && WIFEXITED(*status) && WEXITSTATUS(*status))
					++errors, printf("Test%d: failed: exit status %d\n", i + 1, WEXITSTATUS(*status));
			}
		}

		errors += verify(i, out_name, results[i][0], prog_name, "stdout");
		errors += verify(i, err_name, results[i][1], prog_name, "stderr");
	}

	if (errors)
		printf("%d/%d tests failed\n", errors, tests);
	else
		printf("All tests passed\n");

	exit(0);
}

#endif

/* vi:set ts=4 sw=4: */
